larq_compute_engine/tflite/kernels/quantization.cc

```diff
@@ -0,0 +1,138 @@
+#include "larq_compute_engine/core/bitpack_utils.h"
+#include "tensorflow/lite/c/common.h"
+#include "tensorflow/lite/kernels/internal/cppmath.h"
+#include "tensorflow/lite/kernels/internal/tensor.h"
+#include "tensorflow/lite/kernels/kernel_util.h"
+
+using namespace tflite;
+
+namespace ce = compute_engine;
+
+namespace compute_engine {
+namespace tflite {
+
+using ce::core::TBitpacked;
+
+TfLiteStatus QuantizePrepare(TfLiteContext* context, TfLiteNode* node) {
+  TF_LITE_ENSURE_EQ(context, NumInputs(node), 1);
+  TF_LITE_ENSURE_EQ(context, NumOutputs(node), 1);
+
+  const TfLiteTensor* input = GetInput(context, node, 0);
+  TfLiteTensor* output = GetOutput(context, node, 0);
+
+  TF_LITE_ENSURE(context,
+                 input->type == kTfLiteFloat32 || input->type == kTfLiteInt8);
+  TF_LITE_ENSURE_EQ(context, output->type, kTfLiteInt32);
+
+  TF_LITE_ENSURE_EQ(context, NumDimensions(input), NumDimensions(output));
+
+  int num_dims = NumDimensions(input);
+
+  TfLiteIntArray* output_dims = TfLiteIntArrayCreate(num_dims);
+  // The first n-1 dimensions are equal
+  for (int i = 0; i < num_dims - 1; ++i) {
+    output_dims->data[i] = input->dims->data[i];
+  }
+  // The last dimension is bitpacked
+  output_dims->data[num_dims - 1] =
+      ce::core::GetPackedSize(input->dims->data[num_dims - 1]);
+
+  return context->ResizeTensor(context, output, output_dims);
+}
+
+TfLiteStatus DequantizePrepare(TfLiteContext* context, TfLiteNode* node) {
+  TF_LITE_ENSURE_EQ(context, NumInputs(node), 1);
+  TF_LITE_ENSURE_EQ(context, NumOutputs(node), 1);
+
+  const TfLiteTensor* input = GetInput(context, node, 0);
+  TfLiteTensor* output = GetOutput(context, node, 0);
+
+  TF_LITE_ENSURE_EQ(context, input->type, kTfLiteInt32);
+  TF_LITE_ENSURE(context,
+                 output->type == kTfLiteFloat32 || output->type == kTfLiteInt8);
+
+  TF_LITE_ENSURE_EQ(context, NumDimensions(input), NumDimensions(output));
+
+  int num_dims = NumDimensions(input);
+
+  // The first n-1 dimensions are equal
+  for (int i = 0; i < num_dims - 1; ++i) {
+    TF_LITE_ENSURE_EQ(context, output->dims->data[i], input->dims->data[i]);
+  }
+  // The last dimension is bitpacked
+  int packed_channels = input->dims->data[num_dims - 1];
+  int unpacked_channels = output->dims->data[num_dims - 1];
+  TF_LITE_ENSURE_EQ(context, packed_channels,
+                    ce::core::GetPackedSize(unpacked_channels));
+
+  // We don't support resizing here, because we can not know the number of
+  // output channels based on the number of input channels
```

Comment:
What impact does this have?

Reply:
I think we never use this resizing: basically you can resize the input tensor of your model, for example when you use a different camera resolution. The `Prepare` functions are then all called again (`Init` is not called again), and they are expected to resize their internal buffers (im2col buffer for example) and propagate a new output size, to be sent to the `Prepare` function of the next op.
This way you can use the same `.tflite` file on different input image sizes. Also, I think early toco converters would not store shape information of intermediate tensors in the `.tflite` file, so back then ops definitely had to compute their output size, although I'm not entirely sure about this.
Note that tf-micro does *not* support this resizing, and expects shapes to be present in the `.tflite` file.

For the Dequantize op, it can't compute the number of output channels based on the input shape, since it only knows the packed size. We could add it as a custom option (similar to `input_channels` in `LceBconv2d`), but I thought it'd be easier to just read it from the output tensor, and assume that no resizing will happen. Especially since we never expect to actually use the Dequantize op in a real model.

Reply:
Makes sense, thanks :+1: I guess also even if we somehow were in a scenario that needed resizing, we'd expect the next op to perform validation of it's input tensor size against what it expects, so it wouldn't fail silently.